Source/Core/VideoBackends/Metal/MTLObjectCache.mm

```diff
@@ -0,0 +1,512 @@
+    },
+    [true] = {
+      [static_cast<int>(DstBlendFactor::Zero)]        = MTLBlendFactorZero,
+      [static_cast<int>(DstBlendFactor::One)]         = MTLBlendFactorOne,
+      [static_cast<int>(DstBlendFactor::SrcClr)]      = MTLBlendFactorSourceColor,
+      [static_cast<int>(DstBlendFactor::InvSrcClr)]   = MTLBlendFactorOneMinusSource1Color,
+      [static_cast<int>(DstBlendFactor::SrcAlpha)]    = MTLBlendFactorSource1Alpha,
+      [static_cast<int>(DstBlendFactor::InvSrcAlpha)] = MTLBlendFactorOneMinusSource1Alpha,
+      [static_cast<int>(DstBlendFactor::DstAlpha)]    = MTLBlendFactorDestinationAlpha,
+      [static_cast<int>(DstBlendFactor::InvDstAlpha)] = MTLBlendFactorOneMinusDestinationAlpha,
+    },
+  };
+  // clang-format on
+  return factors[src1][static_cast<int>(factor)];
+}
+
+static MTLBlendFactor Convert(SrcBlendFactor factor, bool src1)
+{
+  // clang-format off
+  static constexpr MTLBlendFactor factors[2][8] = {
+    [false] = {
+      [static_cast<int>(SrcBlendFactor::Zero)]        = MTLBlendFactorZero,
+      [static_cast<int>(SrcBlendFactor::One)]         = MTLBlendFactorOne,
+      [static_cast<int>(SrcBlendFactor::DstClr)]      = MTLBlendFactorDestinationColor,
+      [static_cast<int>(SrcBlendFactor::InvDstClr)]   = MTLBlendFactorOneMinusDestinationColor,
+      [static_cast<int>(SrcBlendFactor::SrcAlpha)]    = MTLBlendFactorSourceAlpha,
+      [static_cast<int>(SrcBlendFactor::InvSrcAlpha)] = MTLBlendFactorOneMinusSourceAlpha,
+      [static_cast<int>(SrcBlendFactor::DstAlpha)]    = MTLBlendFactorDestinationAlpha,
+      [static_cast<int>(SrcBlendFactor::InvDstAlpha)] = MTLBlendFactorOneMinusDestinationAlpha,
+    },
+    [true] = {
+      [static_cast<int>(SrcBlendFactor::Zero)]        = MTLBlendFactorZero,
+      [static_cast<int>(SrcBlendFactor::One)]         = MTLBlendFactorOne,
+      [static_cast<int>(SrcBlendFactor::DstClr)]      = MTLBlendFactorDestinationColor,
+      [static_cast<int>(SrcBlendFactor::InvDstClr)]   = MTLBlendFactorOneMinusDestinationColor,
+      [static_cast<int>(SrcBlendFactor::SrcAlpha)]    = MTLBlendFactorSource1Alpha,
+      [static_cast<int>(SrcBlendFactor::InvSrcAlpha)] = MTLBlendFactorOneMinusSource1Alpha,
+      [static_cast<int>(SrcBlendFactor::DstAlpha)]    = MTLBlendFactorDestinationAlpha,
+      [static_cast<int>(SrcBlendFactor::InvDstAlpha)] = MTLBlendFactorOneMinusDestinationAlpha,
+    },
+  };
+  // clang-format on
+  return factors[src1][static_cast<int>(factor)];
+}
+
+class Metal::ObjectCache::Internal
+{
+public:
+  using StoredPipeline = std::pair<MRCOwned<id<MTLRenderPipelineState>>, PipelineReflection>;
+  /// Holds only the things that are actually used in a Metal pipeline
+  struct PipelineID
+  {
+    struct VertexAttribute
+    {
+      // Just hold the things that might differ while using the same shader
+      // (Really only a thing for ubershaders)
+      u8 offset : 6;
```

Comment:
I guess I just don't see the appeal of `BitField` for internal-use structs.

By switching to `BitField`, I...
- Have to manually calculate the offset of every field
- Have to recalculate those offsets every time I move something
- Can no longer see the values of individual fields in a debugger without explicitly querying them with `expr`
- Can be assured a specific layout that I didn't care about because this struct isn't being serialized and sent to other copies of Dolphin

It just seems like a lot of downsides for very little benefit

Reply:
At least with MSVC, you can view it in the debugger, but that's because we provide a `.natvis` file that formats it automatically.

A major benefit of `BitField` is that some compilers behave in a weird way when you use an `enum class` field with bitfields (in particular older versions of GCC), though it seems that this has [since been fixed](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=61414). That doesn't necessarily apply here, though.